### PR TITLE
remove catboost training dir

### LINKED ISSUE #62: Redirect the catboost_info subfolder created by CatboostEstimator
Feedback from sebhrusen (from the automlbenchmark)
CatboostEstimator is creating and filling a catboost_info subfolder in the running directory. We should be able to pass a 'train_dir' param to Catboost to avoid that.

For example at AutoML level, accept a tmpdir and pass it to each algo supporting an equivalent property (or pass a dedicated subfolder, for example tmpdir/catboost for Catboost and so on).

Reference:
https://github.com/openml/automlbenchmark/pull/270

### PR DESCRIPTION
**Pull request by sonichi**
Use a temporary directory as train_dir and remove it after catboost training

close #62

### CODE REVIEW

File: flaml/model.py [hunk 1]
```diff
@@ -730,16 +732,19 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
-            CatBoostEstimator._smallmodel = self.estimator_class(**self.params)
+            CatBoostEstimator._smallmodel = self.estimator_class(
+                train_dir=train_dir, **self.params)
             CatBoostEstimator._smallmodel.fit(
                 X_train, y_train, cat_features=cat_features, **kwargs)
             CatBoostEstimator._t1 = time.time() - start_time
             if CatBoostEstimator._t1 >= budget:
                 self.params["n_estimators"] = n_iter
                 self._model = CatBoostEstimator._smallmodel
+                shutil.rmtree(train_dir)
```
Comment (qingyun-wu) [inline flaml/model.py:743]: will this trigger permission errors?

Comment (qingyun-wu) [reply flaml/model.py:743]: Add 'try'?

File: flaml/model.py [hunk 2]
```diff
@@ -769,13 +775,14 @@ def fit(self, X_train, y_train, budget=None, **kwargs):
...
             from catboost import Pool
-            model = self.estimator_class(**self.params)
+            model = self.estimator_class(train_dir=train_dir, **self.params)
             model.fit(
                 X_tr, y_tr, cat_features=cat_features,
                 eval_set=Pool(
                     data=X_train[n:], label=y_train[n:],
                     cat_features=cat_features),
                 **kwargs)   # model.get_best_iteration()
+            shutil.rmtree(train_dir)
```
Comment (qingyun-wu) [inline flaml/model.py:785]: same problem as before

File: flaml/automl.py [hunk 3]
```diff
@@ -154,12 +155,16 @@ def _prepare_sample_train_data(self, sample_size):
...
+            if self.groups is not None:
+                groups = self.groups[:sample_size]
         else:
             sampled_X_train = self.X_train_all
             sampled_y_train = self.y_train_all
             if 'sample_weight' in self.fit_kwargs:
                 sampled_weight = self.sample_weight_all
-        return sampled_X_train, sampled_y_train, sampled_weight
+            if self.groups is not None:
+                groups = self.groups_all
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/automl.py:166]: When using holdout, we may take a subset and set `self_groups` for the subset.

**Review by qingyun-wu — APPROVED** [review]

**Review by aminsaied — APPROVED** [review]

**Pull request merged**